oml/retrieval/postprocessors/pairwise.py

```diff
@@ -0,0 +1,206 @@
+            distances += offset.unsqueeze(-1)
+        else:
+            # Pairwise postprocessor has been applied to all possible pairs, so, there are no rest distances.
+            # Thus, we don't need to care about order and offset at all.
+            pass
+
+        distances = assign_2d(x=distances, indices=ii_top.view(n_queries, top_n), new_values=distances_upd)
+
+        assert list(distances.shape) == [n_queries, n_galleries]
+
+        return distances
+
+    def inference(
+        self, queries: Union[List[Path], Tensor], galleries: Union[List[Path], Tensor], ii_top: Tensor, top_n: int
+    ) -> Tensor:
+        """
+        Depends on the exact types of queries/galleries this method may be implemented differently.
+
+        Args:
+            queries: Queries with the length of ``Q``
+            galleries: Galleries with the length of ``G``
+            ii_top: Indices of the closest galleries with the shape of ``[Q, top_n]``
+            top_n: Defines the number of the closest galleries to re-rank
+
+        Returns:
+            Updated distance matrix with the shape of ``[Q, G]``.
+
+        """
+        raise NotImplementedError()
+
+
+class PairwiseEmbeddingsPostprocessor(PairwisePostprocessor):
+    def __init__(
+        self,
+        top_n: int,
+        pairwise_model: IPairwiseDistanceModel,
+        num_workers: int,
+        batch_size: int,
+        verbose: bool = False,
+    ):
+        """
+        Args:
+            top_n: Model will be applied to the ``num_queries * top_n`` pairs formed by each query
+                and ``top_n`` most relevant galleries.
+            pairwise_model: Model which is able to take two embeddings as inputs
+                and estimate the *distance* (not in a strictly mathematical sense) between them.
+            num_workers: Number of workers in DataLoader
+            batch_size: Batch size that will be used in DataLoader
+            verbose: Set ``True`` if you want to see progress bar for an inference
+
+        """
+        assert top_n > 1, "Number of galleries for each query to process has to be greater than 1."
+
+        self.top_n = top_n
+        self.model = pairwise_model
+        self.num_workers = num_workers
+        self.batch_size = batch_size
+        self.verbose = verbose
+
+    def inference(self, queries: Tensor, galleries: Tensor, ii_top: Tensor, top_n: int) -> Tensor:
+        """
+        Args:
+            queries: Queries representations with the length of ``Q``
+            galleries: Galleries representations with the length of ``G``
+            ii_top: Indices of the closest galleries with the shape of ``[Q, top_n]``
+            top_n: Defines the number of the closest galleries to re-rank
+
+        Returns:
+            Updated distance matrix with the shape of ``[Q, G]``.
+
+        """
+        queries = queries.repeat_interleave(top_n, dim=0)
+        galleries = galleries[ii_top]
```

Comment:
If `ii_top` is of size `[Q, top_n]` then `galleries` will be of size `[Q, top_n, dim]`, where `dim` is the embeddings dimension. It doesn't look like it is supposed to be.

Reply:
thx for your findings 
code is okay, but I will fix comments

Reply:
thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`